### PR TITLE
BUG: tz_localize(UTC) not making a copy

### PR DESCRIPTION
**Pull request by jbrockmendel**
- [ ] closes #xxxx (Replace xxxx with the Github issue number)
- [ ] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [ ] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [ ] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

### CODE REVIEW

**Comment by jreback** [comment]
https://dev.azure.com/pandas-dev/pandas/_build/results?buildId=75654&view=logs&j=915dbb23-6529-54ae-059d-ec0b59ba10e8&t=86b091c2-17fe-52fe-2454-5f56a371e155 looks legit

**Comment by jbrockmendel** [comment]
I think thats the test that #46450 fixed

File: pandas/tests/indexes/datetimes/test_timezones.py [hunk 1]
```diff
@@ -327,6 +327,16 @@ def test_tz_convert_unsorted(self, tzstr):
     # -------------------------------------------------------------
     # DatetimeIndex.tz_localize
 
+    def test_tz_localize_utc_copies(self):
+        times = ["2015-03-08 01:00", "2015-03-08 02:00", "2015-03-08 03:00"]
+        index = DatetimeIndex(times)
+
+        res = index.tz_localize("UTC")
```
Comment (mroeschke) [inline pandas/tests/indexes/datetimes/test_timezones.py:334]: Probably good to test with `utc_fixture`.

Also based on the code code `None` also returns a copy now too (and should be tested)?

**Review by mroeschke — COMMENTED** [review]
Warrants a whatsnew?

File: pandas/tests/indexes/datetimes/test_timezones.py (hunk 1 above)
Comment (jbrockmendel) [reply pandas/tests/indexes/datetimes/test_timezones.py:334]: sure.  will use the fixture in the index tests and do the None just once in the tslibs tests

**Comment by mroeschke** [comment]
You may be aware, but looks like this PR didn't fully pass the CI checks

https://dev.azure.com/pandas-dev/pandas/_build/results?buildId=75723&view=logs&jobId=f64eb602-b0ed-5dcc-0a60-a7288ad39597&j=f64eb602-b0ed-5dcc-0a60-a7288ad39597&t=129bd72d-2137-55bd-1811-c4ce50c6dce8

```
_______________________ test_tz_localize_to_utc_copies ________________________
[gw1] win32 -- Python 3.8.12 C:\Miniconda\envs\pandas-dev\python.exe

    def test_tz_localize_to_utc_copies():
        # GH#46460
        arr = np.arange(5, dtype="i8")
        result = tzconversion.tz_convert_from_utc(arr, tz=UTC)
        tm.assert_numpy_array_equal(result, arr)
        assert not np.shares_memory(arr, result)
    
>       result = tzconversion.tz_convert_from_utc(arr, tz=None)

pandas\tests\tslibs\test_conversion.py:60: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
pandas\_libs\tslibs\tzconversion.pyx:506: in pandas._libs.tslibs.tzconversion.tz_convert_from_utc
    converted = _tz_convert_from_utc(vals, tz)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

>   local_val = utc_val + deltas[pos]
E   UnboundLocalError: local variable 'deltas' referenced before assignment

pandas\_libs\tslibs\tzconversion.pyx:571: UnboundLocalError
```

**Comment by jbrockmendel** [comment]
yah, #46478 includes a fix

**Comment by jbrockmendel** [comment]
also #46489 is a targeted fix

**Pull request merged**